File: backend/interfaces/resource/order_books_snapshot.go
package resource

import (
	"hamster/domain/model"
	"time"
)

type OrderBooksSnapshot struct {
	Id                 uint64    `json:"id"`
	Time               time.Time `json:"time"`
	LowestAskPrice     float64   `json:"lowestAskPrice"`
	LowestAskQuantity  float64   `json:"lowestAskQuantity"`
	HighestBidPrice    float64   `json:"highestBidPrice"`
	HighestBidQuantity float64   `json:"highestBidQuantity"`
}

func NewOrderBooksHistory(model *model.OrderBooksSnapshot) *OrderBooksSnapshot {
	return &OrderBooksSnapshot{
		Id:                 model.Id,
		Time:               model.Time,
		LowestAskPrice:     model.LowestAskPrice,
		LowestAskQuantity:  model.LowestAskQuantity,
		HighestBidPrice:    model.HighestBidPrice,
		HighestBidQuantity: model.HighestBidQuantity,
	}
}

File: frontend/model/orderBooksSnapshot.ts
import apiClient from '../lib/apiClient';
import { useQuery } from 'react-query';
import dayjs from 'dayjs';

export class OrderBooksSnapshot {
  unix: number;

  constructor(
    public id: number,
    public time: Date,
    public lowestAskPrice: number,
    public lowestAskQuantity: number,
    public highestBidPrice: number,
    public highestBidQuantity: number,
  ) {
    this.unix = dayjs(time).valueOf();
  }
}

export function useOrderBooksSnapshotsQuery(from: number, to: number) {
  const path = '/order_books/snapshots';
  const key = `${path}:${from}-${to}`;
  return useQuery<OrderBooksSnapshot[]>(
    key,
    async () => {
      const { data } = await apiClient.get(path, {
        params: { from, to },
      });

      return data.map(
        (h: any) =>
          new OrderBooksSnapshot(
            h.id,
            h.time,
            h.lowestAskPrice,
            h.lowestAskQuantity,
            h.highestBidPrice,
            h.highestBidQuantity,
          ),
      );
    },
    { keepPreviousData: true },
  );
}
